ggml.c

```diff
@@ -1958,40 +1958,90 @@ static void ggml_vec_dot_q4_0(const int n, float * restrict s, const void * rest
     // Horizontal sum of all lanes of the accumulator
     sumf = _mm512_reduce_add_ps( acc0 ) + _mm512_reduce_add_ps( acc1 );
 #elif defined(__AVX2__)
+// Input: 32 Nibbles (16 bytes) at *p0 
+// Output: 2 vectors with 16 values of type int16_t 
+#define EXPAND_32_Q4_NIBBLES_INTO_TWO_M256_VECTORS(OUT_HIGH,OUT_LOW,IN_SRC)     \
+        /* get first input */                                                   \
+        /* Load 16 bytes from memory */                                         \
+        const __m128i tmp_##OUT_HIGH =                                          \
+            _mm_loadu_si128( (const __m128i_u *) IN_SRC);                       \
+                                                                                \
+        /* Expand bytes into uint16_t values */                                 \
+        const __m256i bytes_##OUT_HIGH = _mm256_cvtepu8_epi16(tmp_##OUT_HIGH);  \
+                                                                                \
+        /* Unpack values into individual bytes */                               \
+        const __m256i pre_shift_##OUT_HIGH =                                    \
+            _mm256_andnot_si256( lowMask, bytes_##OUT_HIGH );                   \
+        __m256i OUT_HIGH = _mm256_srli_epi16( pre_shift_##OUT_HIGH, 4 );        \
+                                                                                \
+        __m256i OUT_LOW = _mm256_and_si256( lowMask, bytes_##OUT_HIGH );        \
+        /* Now we have a vector with bytes in [ 0 .. 15 ] interval. 
```

Comment:
@sw Thank you for the impulse. I myself was not happy with the macros either. I've converted it into "normal" code. 

The assembler code generated on gcc 11.3 was identical. It would be awesome if someone else could verify.

Reply:
Rebased to master at d8d4e865cd481b18f10508ffee35db903767ef5c